### PR TITLE
feat(debugger): Highlight memory region for any instruction

### PR DESCRIPTION
**Pull request by soyccan**
# Highlight memory region for any instruction in debugger

## Motivation

Currently, debugger highlights the 32-byte memory word that is being read/write in cyan/red when an
MLOAD/MSTORE is encountered, respectively.

But there are quite a few instructions that access memory, e.g. CALL, CALLDATACOPY, SHA3...
This update is to colorize the memory region (which may be of any size) for all of them.

This is an enhancement relevant to #4440. A related work #4681 was to highlight unaligned memory.

## Solution

All instructions that access memory are considered. Memory regions are highlighted in cyan/red for
read/write, consistent with MLOAD/MSTORE. Also, memory regions that are just written by the previous
instruction is highlighted in green. Memory regions may be any length.

image

### CODE REVIEW

File: crates/debugger/src/lib.rs [hunk 1]
```diff
@@ -858,6 +858,50 @@ Line::from(Span::styled("[t]: stack labels | [m]: memory decoding | [shift + j/k
...
+    fn get_memory_access(
+        op: u8,
+        stack: &Vec<U256>,
+    ) -> (Option<usize>, Option<usize>, Option<usize>, Option<usize>) {
+        // The index on the stack that indicates the memory offset/size to be accessed
+        // (read memory offset, read memory size, write memory offset, write memory size)
+        // >= 1: the stack index
+        // 0: no memory access
+        // -1: a fixed size of 32 bytes
+        // -2: a fixed size of 1 byte
```
Comment (mattsse) [inline crates/debugger/src/lib.rs:870]: can we move this to function doc instead of a comment?

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, I like this.

didn't review all the offsets in detail, but I'm willing to try this. changes lgtm and should be easy to fix if something's wrong.

smol doc nit

File: crates/debugger/src/lib.rs [hunk 2]
```diff
@@ -858,6 +858,50 @@ Line::from(Span::styled("[t]: stack labels | [m]: memory decoding | [shift + j/k
         f.render_widget(paragraph, area);
     }
 
+    fn get_memory_access(
```
Comment (Evalir) [inline crates/debugger/src/lib.rs:861]: same comment as matt, lets add function docs (`///`)

File: crates/debugger/src/lib.rs [hunk 3]
```diff
@@ -924,8 +969,11 @@ Line::from(Span::styled("[t]: stack labels | [m]: memory decoding | [shift + j/k
                     .map(|(j, byte)| {
                         Span::styled(
                             format!("{byte:02x} "),
-                            if let (Some(w), Some(color)) = (word, color) {
-                                if (i == w / 32 && j >= w % 32) || (i == w / 32 + 1 && j < w % 32) {
+                            if let (Some(w), Some(size), Some(color)) = (word, size, color) {
+                                if (i == w / 32 && j >= w % 32) ||
+                                    (i > w / 32 && i < (w + size - 1) / 32) ||
+                                    (i == (w + size - 1) / 32 && j <= (w + size - 1) % 32)
```
Comment (Evalir) [inline crates/debugger/src/lib.rs:975]: Can we add some docs here too? actually not super clear on what this does without comments.

**Review by Evalir — CHANGES_REQUESTED** [review]

**Review by mattsse — APPROVED** [review]
lgtm, tysm

File: crates/debugger/src/lib.rs (hunk 3 above)
Comment (soyccan) [reply crates/debugger/src/lib.rs]: Added comments. Now should be clearer.

**Pull request merged**